src/ucs/datastruct/piecewise_func.h

```diff
@@ -0,0 +1,254 @@
+/*
+ * Copyright (c) NVIDIA CORPORATION & AFFILIATES, 2024. ALL RIGHTS RESERVED.
+ *
+ * See file LICENSE for terms.
+ */
+
+#ifndef UCS_PIECEWISE_FUNC_H_
+#define UCS_PIECEWISE_FUNC_H_
+
+#include <ucs/datastruct/linear_func.h>
+#include <ucs/datastruct/list.h>
+#include <ucs/debug/assert.h>
+#include <stddef.h>
+#include <limits.h>
+
+
+/*
+ * A piecewise func segment which represents linear function on a half-open
+ * interval (end is included, start is not).
+ */
+typedef struct {
+    ucs_linear_func_t func;
+    size_t            end;
+    ucs_list_link_t   list;
+} ucs_piecewise_segment_t;
+
+
+/**
+ * A piecewise function consisting of one or more segments.
+ */
+typedef struct {
+    ucs_piecewise_segment_t head_segment;
+    ucs_list_link_t         head;
+} ucs_piecewise_func_t;
+
+
+/**
+ * Intialize a piecewise function which represents f(x) = c + m*x on the
+ * [0, SIZE_MAX] range.
+ *
+ * @param [in] func    Piecewise function to initialize.
+ * @param [in]  c      Piecewise function constant factor.
+ * @param [in]  m      Piecewise function multiplicative factor.
+ *
+ */
+static UCS_F_ALWAYS_INLINE void
+ucs_piecewise_func_init(ucs_piecewise_func_t *func, double c, double m)
+{
+    func->head_segment.func = ucs_linear_func_make(c, m);
+    func->head_segment.end  = SIZE_MAX;
+    ucs_list_head_init(&func->head);
+
+    ucs_list_insert_after(&func->head, &func->head_segment.list);
+}
+
+
+/**
+ * Free a piecewise function allocated memory.
+ *
+ * @param [in] func    Piecewise function to destroy.
+ */
+static UCS_F_ALWAYS_INLINE void
+ucs_piecewise_func_free(ucs_piecewise_func_t *func)
+{
+    ucs_piecewise_segment_t *seg, *tmp;
+
+    ucs_list_for_each_safe(seg, tmp, &func->head, list) {
+        /* No need to deallocate head_segment since it always exists and
+         * allocated statically
+         */
+        if (seg != &func->head_segment) {
+            ucs_free(seg);
+        }
+    }
+}
```

Comment:
What do you mean under `empty the list`?

After this operation there is no more valid elements in the list. We can call `ucs_list_head_init` after freeing the segments to leave the list in the valid state, but the piecewise function still would be in invalid state, since it should always contain at least one segment.

Reply:
ok, sounds good